File: python/problems/leetcode/0234_palindrome_linked_list/solution.py
"""LeetCode 234: Palindrome Linked List."""

from __future__ import annotations

from typing import Optional


class ListNode:
    def __init__(self, val: int = 0, next: Optional["ListNode"] = None) -> None:
        self.val = val
        self.next = next


class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if head is None or head.next is None:
            return True
        slow = fast = head
        while fast is not None and fast.next is not None:
            slow = slow.next  # type: ignore[assignment]
            fast = fast.next.next
        prev: Optional[ListNode] = None
        cur: Optional[ListNode] = slow
        while cur is not None:
            nxt = cur.next
            cur.next = prev
            prev = cur
            cur = nxt
        a: Optional[ListNode] = head
        b: Optional[ListNode] = prev
        while b is not None:
            assert a is not None
            if a.val != b.val:
                return False
            a = a.next
            b = b.next
        return True


def from_list(vs: list[int]) -> Optional[ListNode]:
    dummy = ListNode()
    cur = dummy
    for v in vs:
        cur.next = ListNode(v)
        cur = cur.next
    return dummy.next


if __name__ == "__main__":
    assert Solution().isPalindrome(from_list([1, 2, 2, 1]))
    assert not Solution().isPalindrome(from_list([1, 2]))
    assert Solution().isPalindrome(from_list([1]))
    print("0234 OK")

File: cpp/problems/leetcode/0234_palindrome_linked_list/solution.cpp
// LeetCode 234
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int v) : val(v), next(nullptr) {}
};

class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if (!head || !head->next) return true;
        ListNode *slow = head, *fast = head;
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        ListNode *prev = nullptr, *cur = slow;
        while (cur) {
            ListNode* nxt = cur->next;
            cur->next = prev;
            prev = cur;
            cur = nxt;
        }
        for (ListNode *a = head, *b = prev; b; a = a->next, b = b->next)
            if (a->val != b->val) return false;
        return true;
    }
};

static ListNode* fromList(vector<int> vs) {
    ListNode dummy(0);
    ListNode* cur = &dummy;
    for (int v : vs) {
        cur->next = new ListNode(v);
        cur = cur->next;
    }
    return dummy.next;
}

int main() {
    assert(Solution().isPalindrome(fromList({1, 2, 2, 1})));
    assert(!Solution().isPalindrome(fromList({1, 2})));
    assert(Solution().isPalindrome(fromList({1})));
    cout << "0234 OK" << endl;
    return 0;
}
